pkg/gardenlet/controller/shoot/shoot/cleaner.go

```diff
@@ -0,0 +1,204 @@
+func (c *cleaner) DeleteExtensionObjects(ctx context.Context) error {
+	return utilclient.ApplyToObjectKinds(ctx, func(kind string, objectList client.ObjectList) flow.TaskFn {
+		return func(ctx context.Context) error {
+			c.log.Info("Deleting all extension resources", "kind", kind, "namespace", c.seedNamespace)
+			return extensions.DeleteExtensionObjects(ctx, c.seedClient, objectList, c.seedNamespace, nil)
+		}
+	}, extensionKindToObjectList)
+}
+
+// WaitUntilExtensionObjectsDeleted waits until all extension objects in the shoot namespace have been deleted.
+func (c *cleaner) WaitUntilExtensionObjectsDeleted(ctx context.Context) error {
+	return utilclient.ApplyToObjectKinds(ctx, func(kind string, objectList client.ObjectList) flow.TaskFn {
+		return func(ctx context.Context) error {
+			return extensions.WaitUntilExtensionObjectsDeleted(ctx, c.seedClient, c.log, objectList, kind, c.seedNamespace, defaultInterval, defaultTimeout, nil)
+		}
+	}, extensionKindToObjectList)
+}
+
+// DeleteMCMResources deletes all MachineControllerManager resources in the shoot namespace.
+func (c *cleaner) DeleteMCMResources(ctx context.Context) error {
+	return utilclient.ApplyToObjectKinds(ctx, func(kind string, objectList client.ObjectList) flow.TaskFn {
+		return utilclient.ForceDeleteObjects(ctx, c.log, c.seedClient, kind, c.seedNamespace, objectList)
+	}, mcmKindToObjectList)
+}
+
+// WaitUntilMCMResourcesDeleted waits until all MachineControllerManager resources in the shoot namespace have been deleted.
+func (c *cleaner) WaitUntilMCMResourcesDeleted(ctx context.Context) error {
+	return utilclient.ApplyToObjectKinds(ctx, func(kind string, objectList client.ObjectList) flow.TaskFn {
+		return func(ctx context.Context) error {
+			return kubernetesutils.WaitUntilResourcesDeleted(ctx, c.seedClient, objectList, defaultInterval, client.InNamespace(c.seedNamespace))
+		}
+	}, mcmKindToObjectList)
+}
+
+// DeleteManagedResources removes all remaining finalizers and deletes all ManagedResource resources in the shoot namespace.
+func (c *cleaner) DeleteManagedResources(ctx context.Context) error {
+	c.log.Info("Deleting all ManagedResource resources in namespace", "namespace", c.seedNamespace)
+	if err := c.seedClient.DeleteAllOf(ctx, &resourcesv1alpha1.ManagedResource{}, client.InNamespace(c.seedNamespace)); err != nil {
+		return err
+	}
+
+	return c.finalizeShootManagedResources(ctx, c.seedNamespace)
+}
+
+// WaitUntilManagedResourcesDeleted waits until all ManagedResource resources in the shoot namespace have been deleted.
+func (c *cleaner) WaitUntilManagedResourcesDeleted(ctx context.Context) error {
+	return kubernetesutils.WaitUntilResourcesDeleted(ctx, c.seedClient, &resourcesv1alpha1.ManagedResourceList{}, defaultInterval, client.InNamespace(c.seedNamespace))
+}
+
+// DeleteSecrets removes all remaining finalizers and deletes all secrets in the shoot namespace.
+func (c *cleaner) DeleteSecrets(ctx context.Context) error {
```

Comment:
Don't we also need to delete `ConfigMap`s? I think the terraformer is creating `ConfigMap`s with a finalizer that could potentially remain?